### ISSUE TITLE
type="email" is not being validated on the server

### ISSUE THREAD

**Issue by andrecasal** [brophdawg11/remix-validity-state#40]
Simple newsletter subscription example:

```tsx
import { Form } from '@remix-run/react'
import { validateServerFormData, InputDefinition } from 'remix-validity-state'
import { json, TypedResponse } from '@remix-run/node'
import { DataFunctionArgs } from '@remix-run/node'

type ActionData = {
	fields?: {
		email: string
	}
	formError?: string
}

const badRequest = (data: ActionData): TypedResponse<ActionData> => json(data, { status: 400 })

type FormSchema = {
	inputs: {
		email: InputDefinition
	}
}

const formDefinition: FormSchema = {
	inputs: {
		email: {
			validationAttrs: {
				type: 'email',
				required: true,
			}
		}
	}
}

export const action: ActionFunction = async ({ request }) => {
	const { valid, submittedValues, inputs } = await validateServerFormData(await request.formData(), formDefinition)
	if (!valid) {
		return badRequest({ formError: `Form not submitted correctly` })
	}

	// All ok, use submittedValues
}

const Newsletter = () => {
	return <Form method="post">
		<h2>Subscribe!</h2>
		<p>Don't miss any of the action!</p>
		<fieldset>
			<input {...{
				...formDefinition.inputs.email.validationAttrs,
				type: 'text' // 👈 Replace type="email" with type="text" to simulate user tampering with UI
			}} name="email" placeholder="you@example.com" />
			<button type="submit">Subscribe</button>
		</fieldset>
	</Form>
}

export default Newsletter
```

Submit an invalid email and notice the `type="email"` validation fails on the server.

**Comment by brophdawg11** [comment]
Yeah this is currently a "by design" case since I'm not sure what exactly the browsers are all doing internally for their `type` validations.  Client-side we just let the browser tell us when an input is valid.  Server-side we have to re-implement the same checks (see `builtInValidations` in the library code).  Things like `required` and `minLength` are simple to re-implement, but I've not yet bothered with the `type` validations.  

I'm definitely not interested in some nasty [regex](https://stackoverflow.com/a/46181) to try to validate email addresses.  Mostly because they're horrible to debug and generally gives tons of false negatives.  I'd definitely also recommend giving [this post](https://medium.com/hackernoon/the-100-correct-way-to-validate-email-addresses-7c4818f24643) a read - this changed how I thought about email addresses significantly many years ago.

I think I could be convinced to do something _very_ simple on the server such as `/.+@.+/` to make sure there's something before and after an `@` character but I doubt I'd want to go any further.  I'll also need to look into what that would mean for the rest of the inputs (type url, tel, date, etc.) since if I do one I'd prefer to do them all and unsure at the moment if that opens up a can of worms.

For now, I'd recommend doing your own simple regex-based check.  In the end though, the proper way to validate an email address on the server is to send it an email with a verification link :)

**Comment by CSFlorin** [comment]
Here's the standard browsers adhere to along with a regex that satisfies it: https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address

**Comment by brophdawg11** [comment]
Oh cool I didn't know the spec gave their own regex, that's handy.  I'll do some digging on the rest of the [types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) as well.  I hope (but doubt) they'll all be so easy 🤣

**Comment by andrecasal** [comment]
Oh I thought this was already being done in a previous version. I'd be happy with a simple `/.+@.+/`.

If I'm not mistaken, there are only two `type`'s that have implicit validation: `email` and `url`.

**Comment by brophdawg11** [comment]
ok cool.  yeah the URL spec is nasty but since node has `URL` available globally, I think we can get by with `try { new URL(value); } catch (e) { return false }` on the server.  If folks are running in non-node environments, they'll need URL available which should be fine for most alternative server-worker-API based environments.  

Thanks for the references!

**Comment by brophdawg11** [comment]
Released in `0.10.0`
